### PR TITLE
Add first version of KEP and some instances

### PR DESCRIPTION
**Pull request by marco-novaes98**

### CODE REVIEW

File: classic_cp/kidneyexchange/IOmanager.jl [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    for i = 2:numberOfPairs+1
+        if isempty(lines[i])
+            compabilities[i-1] = []
+        else
+            line = split(lines[i],' ')
+            compabilities[i-1] = map(x->parse(Int,x), line)
+        end
+    end
+    return InputData(numberOfPairs, density, compabilities)
+end
```
Comment (3rdCore) [inline classic_cp/kidneyexchange/IOmanager.jl:20]: Some comments regarding the input format would be appreciated. Don't forget to add a header to each of your functions to make it easier for future collaborators to work on your code !

File: classic_cp/kidneyexchange/kidneyexchange.jl [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+2 4    #pair 1 can receive a kidney from pairs 2 and 4
+1      #pair 2 can receive a kidney from pair 1
+4      #pair 3 can receive a kidney from pair 4
+3      #pair 4 can receive a kidney from pair 3
+"""
+
+struct InputData
+    numberOfPairs         :: Int
+    density               :: Float16
+    compatibilities       :: Vector{Union{Vector{Int}, Nothing}}
```
Comment (3rdCore) [inline classic_cp/kidneyexchange/kidneyexchange.jl:15]: what does a `Nothing` means ?

File: classic_cp/kidneyexchange/kidneyexchange.jl [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    for i in 1:n
+        if isempty(c[i])
+            push!(pairsEquivalence, nothing)
+        else
+            push!(pairsEquivalence, count)
+            count += 1
+        end
+    end
+
+    c = filter(e -> !isempty(e), c)
```
Comment (3rdCore) [inline classic_cp/kidneyexchange/kidneyexchange.jl:58]: I guess this part was meant to remove all Pairs that could neither give nor receive in any case ?

File: classic_cp/kidneyexchange/kidneyexchange.jl [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    ### Objective ###
+    numberOfExchanges = SeaPearl.IntVar(-n, 0, "numberOfExchanges", trailer) 
+    SeaPearl.addVariable!(model, numberOfExchanges)
+    vars = SeaPearl.AbstractIntVar[]
+    for i in 1:n
+        vars = cat(vars, int2bool(x_offset[i]); dims=1) #TODO int2bool?
+    end
+    push!(vars, numberOfExchanges)
+    objective = SeaPearl.SumToZero(vars, trailer)
+    push!(model.constraints, objective)
```
Comment (3rdCore) [inline classic_cp/kidneyexchange/kidneyexchange.jl:100]: you should not use `push!(model.constraints, your_constraint )` to add a constraint but rather `SeaPearl.addConstraint!`

**Review by 3rdCore — COMMENTED** [review]
Don't forget to add headers to your functions!

File: classic_cp/kidneyexchange/kidneyexchange.jl [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    n = current_n
+    
+    ### Variable declaration ###    
+    x = Vector{SeaPearl.AbstractIntVar}(undef, n) 
+    x_offset = Vector{SeaPearl.AbstractIntVar}(undef, n) 
+
+    for i = 1:n
+        x[i] = SeaPearl.IntVar(1, n, "x_"*string(i), model.trailer)
+        SeaPearl.addVariable!(model, x[i]; branchable=true)
+        x_offset[i] = SeaPearl.IntVarViewOffset(x[i], -i, x[i].id*"-"*string(i))
```
Comment (3rdCore) [inline classic_cp/kidneyexchange/kidneyexchange.jl:76]: You should add them to your model!

File: classic_cp/kidneyexchange/kidneyexchange.jl (hunk 2 above)
Comment (marco-novaes98) [reply classic_cp/kidneyexchange/kidneyexchange.jl]: There is no reason to use "nothing", I has just corrected it

File: classic_cp/kidneyexchange/kidneyexchange.jl (hunk 3 above)
Comment (marco-novaes98) [reply classic_cp/kidneyexchange/kidneyexchange.jl:50]: Exactly, I will add some comments to clarify the instance reduction

Comment (marco-novaes98) [reply classic_cp/kidneyexchange/kidneyexchange.jl:48]: done

File: classic_cp/kidneyexchange/kidneyexchange.jl (hunk 4 above)
Comment (marco-novaes98) [reply classic_cp/kidneyexchange/kidneyexchange.jl]: done

File: classic_cp/kidneyexchange/kidneyexchange.jl [hunk 6]
```diff
@@ -8,98 +8,124 @@ end
...
+        #Fix xEqualIndex_Bool
+        SeaPearl.addConstraint!(model, SeaPearl.ReifiedInSet(x[i], index[i], xEqualIndex_Bool[i], trailer))
+
+        #Fix xNotEqualIndex_Int
+        SeaPearl.addConstraint!(model, SeaPearl.ReifiedInSet(xNotEqualIndex_Int[i], zeroSet, xEqualIndex_Bool[i], trailer))
+
+        #Add incompatibilities to the model
         for j in 1:n
-            if j ∉ c[i]
+            if j ∉ c[i] && i !=j
```
Comment (3rdCore) [inline classic_cp/kidneyexchange/kidneyexchange.jl:93]: So we do not consider the case where a pair can be paired to itself?

Comment (3rdCore) [reply classic_cp/kidneyexchange/kidneyexchange.jl:93]: Sorry, I understood later the underlying process using the `xEqualIndex_Bool` variables.

File: classic_cp/kidneyexchange/kidneyexchange.jl [hunk 7]
```diff
@@ -8,98 +8,124 @@ end
...
+
+    #SeaPearl's solver minimize the objective variable, so we use minusNumberOfExchanges in order to maximize the number of exchanges
+    minusNumberOfExchanges = SeaPearl.IntVar(-n, 0, "minusNumberOfExchanges", trailer) 
+    SeaPearl.addVariable!(model, minusNumberOfExchanges; branchable=false)
     vars = SeaPearl.AbstractIntVar[]
+
+    #Concatenate all values of xNotEqualIndex_Int and minusNumberOfExchanges
     for i in 1:n
-        vars = cat(vars, int2bool(x_offset[i]); dims=1) #TODO int2bool?
+        vars = cat(vars, xNotEqualIndex_Int[i]; dims=1)
```
Comment (3rdCore) [inline classic_cp/kidneyexchange/kidneyexchange.jl:111]: Just to be sure, we are trying to maximize the number of exchanges (ie. the number of times a peer is not linked to itself, is that right ?

File: classic_cp/kidneyexchange/kidneyexchange.jl [hunk 8]
```diff
@@ -8,98 +8,124 @@ end
...
+        xNotEqualIndex_Int[i] = SeaPearl.IntVar(0, 1,"xNotEqualIndex_Int_"*string(i), model.trailer)
     end
 
     ### Constraints ###
     for i = 1:n
+        #Fix xEqualIndex_Bool
+        SeaPearl.addConstraint!(model, SeaPearl.ReifiedInSet(x[i], index[i], xEqualIndex_Bool[i], trailer))
+
+        #Fix xNotEqualIndex_Int
+        SeaPearl.addConstraint!(model, SeaPearl.ReifiedInSet(xNotEqualIndex_Int[i], zeroSet, xEqualIndex_Bool[i], trailer))
```
Comment (3rdCore) [inline classic_cp/kidneyexchange/kidneyexchange.jl:89]: Just to be sure : 
if `xEqualIndex_Bool[i] == True` then : `xNotEqualIndex_Int[i]` = 0 and `x[i] = index[i] = i` 

But the comment line 63 tells that `xNotEqualIndex_Int[i] = 0 => i ∈ cycle`, but `x[i] = index[i] = i` implies that `i ∉ cycle`, which is not possible.what did I do wrong ?

**Review by 3rdCore — APPROVED** [review]
Looks good for me, very nice job! 
Once you answer my question, you can directly squash and merge the PR.

File: classic_cp/kidneyexchange/kidneyexchange.jl [hunk 9]
```diff
@@ -8,98 +8,124 @@ end
...
     
     ### Variable declaration ###    
+
+    # x[i] = j => pair j receive a kidney from pair i
+    # x[i] = i => pair i ∉ cycles
     x = Vector{SeaPearl.AbstractIntVar}(undef, n) 
-    x_offset = Vector{SeaPearl.AbstractIntVar}(undef, n) 
+
+    # "fake" variable to know if a pair participate in a cycle
+    index = Vector{SeaPearl.AbstractVar}(undef, n)
```
[the remark this answers is not in the document]
Comment (marco-novaes98) [reply classic_cp/kidneyexchange/kidneyexchange.jl:54]: The vector model does not allow self-compatible pairs, as there will be no way to distinguish a self-compatible pair (size one cycle) from a pair that does not participate in any cycle. I forget to clarify this in the headers/comments, I will add some explanations.

File: classic_cp/kidneyexchange/kidneyexchange.jl (hunk 7 above)
Comment (marco-novaes98) [reply classic_cp/kidneyexchange/kidneyexchange.jl:111]: Yes, knowing that a peer linked to itself means that the peer does not participate in any cycle (in other words, the patient does not receive a kidney)

File: classic_cp/kidneyexchange/kidneyexchange.jl (hunk 8 above)
Comment (marco-novaes98) [reply classic_cp/kidneyexchange/kidneyexchange.jl:89]: Oh I did a mistake in the comments line 62 and 63 (swap ∉ / ∈). That explains the confusion ^^ 
I will correct that.

**Pull request merged**